Fallback components in generated formulas

ela-kotulska-frequenz opened the pull request:
Fallback components are used in generated formulas. If primary components is unavailable, formula will generate metric from fallback components. Fallback formulas are implemented for:
  - PVPowerFormula
  - ProducerPowerFormula
  - BatteryPowerFormula
  - ConsumerPowerFormula
  - GridPowerFormula
 
All necessary formulas are implemented in this PR

ela-kotulska-frequenz committed:
Add method to get primary and fallback components

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Create fallback formulas for the PvPool.power

Change PVPowerFormula to get primary and fallback of the given
components. For the fallback comonents sub formulas are created
This makes sure that both main and fallback formulas have the
same behaviour and less complexity.

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Add FallbackMetricFetcher to MetricFetcher

FallbackMericFetcher stores fallback formula but generates
and subscribes for it on demand.

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Fix tests after previous commit

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Change receiver limit in mock_resampler

This receiver is only to forward messages to subscribers.
It doesn't need more then 1 last message.
No tests are broken after this.

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Test pv_power formula with fallback components

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Update release notes

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

Marenz commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_formula_generator.py:
```diff
@@ -140,3 +140,60 @@ def generate(
+        """Get primary and fallback components within a given set of components.
+
+        This function assumes that fallback component has exactly one meter
+        in from of him.
```
> has exactly one meter in from of him.

..what?

Marenz replied:
Ooh, one *Meter*, not the distance.  This whole method description was very confusing to me..

llucax replied:
"in FRONT of him", that probably the main source of confusion.

Marenz replied:
I actually didn't even notice the missing "front"  XD for me the confusion was more "him.. who is he?". 
Maybe writing "metering component" or so would make it more clear

ela-kotulska-frequenz replied:
Function description was changed.

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_formula_generator.py:
```diff
@@ -140,3 +140,60 @@ def generate(
+        the method looks for its main component, usually a meter, and pairs them
+        together.
+        If a component has no fallback, (e.g. an inverter that has no meter attached),
+        the method simply returns an empty set for that component.
```
I'm a bit confused about this, in the case of an inverter and a meter, shouldn't the inverter be the fallback for the meter? This paragraph seems to imply is the other way around.

ela-kotulska-frequenz replied:
Changed description. Please let me know if current one is better

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_formula_generator.py:
```diff
@@ -140,3 +140,60 @@ def generate(
+        the method simply returns an empty set for that component.
+
+        Args:
+            components: The component to be analyzed.
```
It takes the primary component? I think this methods docs need to be improved a bit and define what is a primary component, what is a fallback component and specify which is being referred to when talking about a component.

ela-kotulska-frequenz replied:
Changed method description. Please let me know if current one is better

llucax replied:
Yeah, thanks! Much clearer now!

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_engine.py:
```diff
@@ -747,6 +748,7 @@ def push_metric(
         data_stream: Receiver[Sample[QuantityT]],
         *,
         nones_are_zeros: bool,
+        fallback_metric_fetcher: FallbackMetricFetcher[QuantityT] | None = None,
```
Super minor suggestion: I would call this just `fallback`, as it is already known from the type it is a metric fetcher.

ela-kotulska-frequenz replied:
done

ela-kotulska-frequenz committed:
Fix MockMicrogrid add_chps method to can create chp without meter

Previously if no_meter was True, meter was created but
didn't send values. Now no meter is created.
No tests are failing after this change

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Implement fallback feature for ProducerPowerFormula

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_steps.py:
```diff
@@ -343,6 +347,41 @@ def apply(self, eval_stack: list[float]) -> None:
         eval_stack.append(val)
 
 
+class FallbackMetricFetcher(ABC, Generic[QuantityT]):
```
For new classes I recommend inheriting directly from `BackgroundService`.

ela-kotulska-frequenz replied:
Wow I didn't know we have this class, thanks :) 
Fixed

llucax replied:
Yeah, and it is about to be replaced by a new one already :laughing: 

* https://frequenz-floss.github.io/frequenz-core-python/latest/reference/frequenz/core/asyncio/#frequenz.core.asyncio.Service
*https://frequenz-floss.github.io/frequenz-core-python/latest/reference/frequenz/core/asyncio/#frequenz.core.asyncio.ServiceBase

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_steps.py:
```diff
@@ -382,6 +428,9 @@ def stream_name(self) -> str:
         """
         return str(self._stream.__doc__)
 
+    def _value_is_invalid(self, value: QuantityT | None) -> bool:
+        return value is None or value.isnan() or value.isinf()
```
I suggest using the positive form instead of negative, as a double negative could be hard to read (`if not self._value_is_invalid(...)` (and structure the method a bit more like a question), so:

```suggestion
    def _is_value_valid(self, value: QuantityT | None) -> bool:
        return value is not None and not value.isnan() and not value.isinf()
```

ela-kotulska-frequenz replied:
done

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_steps.py:
```diff
@@ -390,7 +439,41 @@ async def fetch_next(self) -> Sample[QuantityT] | None:
+                self._next_value = result[0]
+                # We can stop the fallback metric fetcher if we have a valid value from
+                # primary component.
+                # This is not tested yet.
```
Probably not to put in a code comment? (the "not tested yet")

shsms replied:
Probably best to drop the stop method completely and create an issue about it instead.

ela-kotulska-frequenz replied:
Done :) 
https://github.com/frequenz-floss/frequenz-sdk-python/issues/1045

llucax reviewed (commented):
I just gave this a quick look. Will leave it for people more familiar with the formula engine.

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_formula_generator.py:
```diff
@@ -42,13 +44,75 @@ class ComponentNotFound(FormulaGenerationError):
 """
 
 
+class FallbackFormulaMetricFetcher(FallbackMetricFetcher[QuantityT]):
```
I'm guessing this is done because of circular imports?  If so adding a comment would be helpful.

Still maybe it would be nicer to put this new class in a separate file like `_fallback_components.py` rather than in `_formula_generator.py`

ela-kotulska-frequenz replied:
> I'm guessing this is done because of circular imports? 

Yes :) Comment was added.

Class was moved to separate file : src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_fallback_formula_metric_fetcher.py

shsms reviewed (commented):
The logic looks good to me.

Would be nice if you put the docstrings and comments through a grammar checker like https://languagetool.org/ or https://www.grammarly.com/grammar-check.

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_fallback_formula_metric_fetcher.py:
```diff
@@ -0,0 +1,55 @@
+        """Check whether the formula engine is running."""
+        return self._receiver is not None
+
+    def start(self) -> None:
+        """Initialize formula engine and start fetching samples."""
+        self._formula_engine = self._formula_generator.generate()  # type: ignore[assignment]
+        assert isinstance(self._formula_engine, FormulaEngine)
+        self._receiver = self._formula_engine.new_receiver()
```
Mmm, interesting, but with this implementation `stop()` and `cancel()` will not work. I guess `cancel()` could still be a NOP, but you should probably override `stop()`.

ela-kotulska-frequenz replied:
`stop()` should work because it has :
```
        if not self._tasks:
            return
```

llucax replied:
OK, I guess I reviewed this too quickly and didn't realized the missing `stop()` in the comment was this stop. I also missed that even when it had a `run()` method it didn't actually have any task running. I think I saw a `run()` and `stop()` and immediately assumed it had some task running.

I think this class is actually a good match for the new abstract `Service` interface, and a good case for splitting the `ServiceBase` with a basic implementation using tasks, as it doesn't really help in this case but it is nice to still have the common `Service` interface with start/stop/is_running/etc. for consistency.

ela-kotulska-frequenz replied:
Ok, I will do that later today (or on Monday).

ela-kotulska-frequenz replied:
sdk doesn't use frequenz-core-python now. Should I add it and use `Service`?
Or should I do that in separate PR?

ela-kotulska-frequenz replied:
frequenz-core-python has no release, yet.
Should I add dependency to the v1.x.x? 

Maybe I could do that in separate PR (this one is pretty big now)?

llucax replied:
No, sorry, I didn't mean to do it now, but I actually got thinking about this, and I wonder why isn't this just a `Receiver` with a couple of extra methods and we are already using `stop()` for some receivers (like in [`Merger`](https://github.com/frequenz-floss/frequenz-channels-python/blob/d8bafe8ee30f661ebe2b420c5fd9b240de049dd7/src/frequenz/channels/_merge.py#L131-L136).

Part of the reasons of splitting this to `core` is so we can use `core` in `channels` too, and maybe make these `Receiver`s a `Service` too.

So my suggestion would be to make it a `Receiver` (so `fetch_next()` would just be `receive()`) and follow the interface of `Service` so in the future when this can inherit from `Service` it is not a breaking change.

**OR**, you can leave things are they are and we take care of it when we introduce the dependency to `core`, which will probably need adapting many other classes :)

ela-kotulska-frequenz replied:
This can't be Receiver, because stopping Receiver won't stop formula.
Here i would like to start/stop formula when needed. This means: subscribe/unsubscribe from all formula components, too.

llucax replied:
I still don't get this.

So there is no `stop()` for now (or there is one inherited from `BackgroundService` that doesn't do anything).

I guess in the future you want to add a `stop()`. You are saying this `stop()` should not stop the `self._formula_engine`? Or the `self._receiver`? Or both? In any case, if you make `FallbackMetricFetcher` a `Receiver`, why can't you just make `stop()` do whatever it should do? If it is a semantic issue, maybe you could call it `pause()` instead of `stop()` to better convey that it is not entirely stopped but just not producing values anymore temporarily?

I think it was also mentioned before, I find it confusing that we have this extra inheritance `FallbackFormulaMetricFetcher` -> `FallbackMetricFetcher`. Can't we get rid of `FallbackMetricFetcher`? What is it for? We already have some unnecessary inheritance in the SDK from the past, I would prefer not to introduce more unless it is really necessary.

ela-kotulska-frequenz replied:
Done, I made it Receiver.

ela-kotulska-frequenz committed:
Remove redundant parentheses in BatteryPowerFormula

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Add fallback formula feature for BatteryPowerFormula

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Fix tests after previous commit

And fix graph in test_batter_pool_power_two_batteries_per_inverter,
Previously meters were connected to the grid and had no successors,
inverters were connected to grid and had no meter.
Now meters are connected to the inverters.

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Add tests for BatteryPowerFormula fallback

Fix mockMicrogrid: create senders always in the same order.
Previous solution was working because we had simple graphs.

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

shsms commented on tests/timeseries/_battery_pool/test_battery_pool.py:
```diff
@@ -590,6 +590,96 @@ async def test_batter_pool_power_two_batteries_per_inverter(
+    mockgrid = MockMicrogrid(
+        graph=gen.to_graph(
+            [
+                ComponentCategory.METER,  # Grid meter - shouldn't be included in formula
```
I think this is not a grid meter.  A grid meter would look like:

```python
    gen.to_graph(
        (
            ComponentCategory.METER,  # Grid meter
            [
                (ComponentCategory.METER, [...]),  # meter with 2 inverters
                (...),  # inverter without meter
            ],
        )
    )
```

Just changing the comment should be enough.

ela-kotulska-frequenz replied:
Fixed, thanks! 
I changed it to be grid meter.

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_formula_generator.py:
```diff
@@ -140,3 +140,77 @@ def generate(
+
+                if (
+                    all(c.category == ComponentCategory.CHP for c in successors)
+                    or all(c.category == ComponentCategory.INVERTER for c in successors)
```
Here you'd need to check if they're all the same type as well.  Either all battery inverters or all PV inverters.

ela-kotulska-frequenz replied:
Right, thanks! The first case is also wrong. I am fixing it now.

ela-kotulska-frequenz replied:
Done

shsms reviewed (commented):
Still looking into it, but got a couple of items.

ela-kotulska-frequenz commented:
first commit: `_get_metric_fallback_components` method is wrong, fixing it now.

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_formula_generator.py:
```diff
@@ -140,3 +140,77 @@ def generate(
+                if len(predecessors) == 1:
+                    predecessor = predecessors.pop()
+
+                    if predecessor.category == ComponentCategory.METER:
```
Not sure if this is the same issue you detected as well, but it could detect grid meters as ev meters, etc. in some cases.

That issue was else where as well, in the component graph.  I made a fix here: https://github.com/frequenz-floss/frequenz-sdk-python/pull/1052

You could just use the new `graph.is_grid_meter()` method to avoid selecting them here.

ela-kotulska-frequenz replied:
Yes, this is what I found. 
Fixed.

shsms commented:
I haven't finished reviewing this yet, but I'm happy to take this in and continue review afterwards because I have to work on these parts anyway.

@llucax if you're not opposed to that, please go ahead and approve.  If there are specific things that you would like to change, please create issues.

I will look into improving the documentation a bit as well, after this is merged.  It is too hard to do that in code review.

shsms commented:
Or maybe better to take this into a feature branch.  Then I can take my fallback values for power distribution PR in there as well, without having to wait for formula support.

And take it in once it all looks good for PVs at least.

ela-kotulska-frequenz committed:
Add fallback formula feature to ConsumerPowerFormula

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Add fallback formula feature to GridPowerFormula

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_producer_power_formula.py:
```diff
@@ -43,35 +50,111 @@ def generate(  # noqa: DOC502
-                "Unable to find any producer components in the component graph. "
-                "Subscribing to the resampling actor with a non-existing "
-                "component id, so that `0` values are sent from the formula."
+        if self._config.component_ids is None:
```
why do we need to check this for producer power formula?

ela-kotulska-frequenz replied:
simplified in last commit

shsms commented on tests/timeseries/_battery_pool/test_battery_pool.py:
```diff
@@ -590,6 +590,99 @@ async def test_batter_pool_power_two_batteries_per_inverter(
         assert (await power_receiver.receive()).value == Power.from_watts(-2.0)
 
 
+async def test_batter_pool_power_fallback_formula(
```
```suggestion
async def test_battery_pool_power_fallback_formula(

ela-kotulska-frequenz replied:
done

ela-kotulska-frequenz committed:
Use SimplePowerFormula in ProducerPowerFormula

This should simplify code

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_steps.py:
```diff
@@ -390,7 +421,37 @@ async def fetch_next(self) -> Sample[QuantityT] | None:
-        self._next_value = await self._stream.receive()
+        if self._fallback is not None and self._fallback.is_running:
+            tasks = [self._stream.receive(), self._fallback.fetch_next()]
+            result = await asyncio.gather(*tasks, return_exceptions=False)
```
Wouldn't it make sense here to actually return exceptions instead of bubbling exceptions up? If one of the two doesn't raise an exception, you can still use that value (and probably log the error), right?

ela-kotulska-frequenz replied:
Done

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_fallback_formula_metric_fetcher.py:
```diff
@@ -0,0 +1,54 @@
+
+    def start(self) -> None:
+        """Initialize the formula engine and start fetching samples."""
+        self._formula_engine = self._formula_generator.generate()  # type: ignore[assignment]
+        assert isinstance(self._formula_engine, FormulaEngine)
```
I'm sorry if I already asked this and got lost by some rebase, but why the `ignore` and `assert` here?

ela-kotulska-frequenz replied:
Generics hell:
 
`generate` method definition:
```
    @abstractmethod
    def generate(
        self,
    ) -> FormulaEngine[QuantityT] | FormulaEngine3Phase[QuantityT]:
        """Generate a formula engine, based on the component graph."""
```

It can return `FormulaEngine` or `FormulaEngine3Phase`. 
* FormulaEngine returns formula output in type `Sample`
* FormulaEngine3Phase returns output as in type `Sample3Phase`

But then `MetricFetcher` works only on FormulaEngine - because it stores receiver that returns `Sample`
FallbackMetricFetcher also has to work on  FormulaEngine - because its output is returned by MetricFetcher.

To solve it in `proper` way we would need separate base class for FormulaEngine and FormulaEngine3Phase.

llucax replied:
I see, yeah, I also bumped into these issues in #822.

Maybe then the `ignore` won't be needed if you do this instead?

```py
        engine = self._formula_generator.generate()
        # We need this assert because generate() can return a FormulaEngine
        # or FormulaEngine3Phase, but in this case we know it will return a
        # FormulaEngine. This helps to silence `mypy` and also to verify our
        # assumptions are still true at runtime
        assert isinstance(engine, FormulaEngine)
        self._formula_engine = engine
```

shsms replied:
Maybe nicer solution is to have `generate` and `generate_3_phase`, and different formulas implement different functions.  But not for this PR.

llucax replied:
* https://github.com/frequenz-floss/frequenz-sdk-python/issues/1055

llucax replied:
@ela-kotulska-frequenz maybe you can add a link to the issue in the comment too :innocent:

llucax reviewed (commented):
A few more minor comments.

llucax commented:
> @llucax if you're not opposed to that, please go ahead and approve. If there are specific things that you would like to change, please create issues.
----
> Or maybe better to take this into a feature branch. Then I can take my fallback values for power distribution PR in there as well, without having to wait for formula support.
> 
> And take it in once it all looks good for PVs at least.
----
OK, I´m not sure what's the current approach, but I'm also fine with take it as it is and create issues if there is more work coming up in this code and it can unblock progress.

Just let me know or approve, and I can create any issues for stuff that I consider relevant after it is merged.

shsms commented:
Could you update the description to say what is done and what is pending?  It seems outdated.

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_steps.py:
```diff
@@ -390,7 +421,37 @@ async def fetch_next(self) -> Sample[QuantityT] | None:
+            result = await asyncio.gather(*tasks, return_exceptions=False)
+
+            # Synchronize the timestamps from two sources
+            while result[0].timestamp < result[1].timestamp:
+                result[0] = await self._stream.receive()
+
+            while result[0].timestamp > result[1].timestamp:
+                result[1] = await self._fallback.fetch_next()
```
I feel like the first check shouldn't be there.  The fallback should be updated to match the original stream, not the other way around.  Else we could skip data.

If the fallback stream is ahead, then we shouldn't fetch from it until the original stream has caught up, and send the values from the original stream until then.

I guess this would also mean that you can't pull from both the streams together using a `asyncio.gather` either.

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_pv_power_formula.py:
```diff
@@ -58,18 +63,79 @@ def generate(  # noqa: DOC502
+                if idx > 0:
+                    builder.push_oper("+")
+
+                # should only be the case if the component is not a meter
+                builder.push_component_metric(
+                    primary_component.component_id,
+                    nones_are_zeros=False,
+                    fallback=fallback_formula,
+                )
+        else:
+            for idx, component in enumerate(pv_components):
+                if idx > 0:
+                    builder.push_oper("+")
+
+                # should only be the case if the component is not a meter
+                builder.push_component_metric(
+                    component.component_id,
+                    nones_are_zeros=component.category != ComponentCategory.METER,
+                )
```
Looks like those are outdated comment lines and don't seem to make sense.  Maybe they can go away.

ela-kotulska-frequenz replied:
Done

shsms commented on tests/timeseries/mock_resampler.py:
```diff
@@ -61,7 +61,7 @@ def power_senders(
                     self._channel_registry.get_or_create(
                         Sample[Quantity], name
-                    ).new_receiver()
+                    ).new_receiver(limit=1)
```
This doesn't feel right, because we want the raw data to queue up if we're not consuming continuously.  What was going wrong without this change?

ela-kotulska-frequenz replied:
Yes, it is not needed anymore.
It was needed before( When we had no synchronization between MetricFetcher and FallbackMetricFetcher).
Removed

shsms replied:
This is still pending.

shsms commented on tests/timeseries/test_logical_meter.py:
```diff
@@ -77,3 +77,69 @@ async def test_pv_power_no_pv_components(self, mocker: MockerFixture) -> None:
+                mockgrid.mock_resampler.next_ts()
+
+                result = await pv_power_receiver.receive()
+                assert result.value == expected_power, (
```
```suggestion
                assert result.value.isclose(expected_power), (

ela-kotulska-frequenz replied:
Why? Those values should be equal.

shsms replied:
This is because of float precision issues, in float arithmatic results.  We had issues of this sort and decided to only use `isclose` from now on, although we still have some old equal checks which need to be migrated.

`isclose` is an equality check in practice, because the default tolerance is very small.

shsms replied:
This is still pending.

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_pv_power_formula.py:
```diff
@@ -58,18 +63,79 @@ def generate(  # noqa: DOC502
+                    builder.push_oper("+")
+
+                # should only be the case if the component is not a meter
+                builder.push_component_metric(
+                    primary_component.component_id,
+                    nones_are_zeros=False,
+                    fallback=fallback_formula,
+                )
```
No reason to think the primary component would always be a meter.

```suggestion
                builder.push_component_metric(
                    primary_component.component_id,
                    nones_are_zeros=primary_component.category != ComponentCategory.METER,
                    fallback=fallback_formula,
                )

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py:
```diff
@@ -65,39 +69,106 @@ def generate(
 
-        battery_inverters = frozenset(
-            frozenset(
+        # all_connected_batteries has to be set because
+        # multiple inverters can be connected to the same battery
```
It has a different meaning otherwise, which is confusing.
```suggestion
        # all_connected_batteries has to be a `set` because
        # multiple inverters can be connected to the same battery

ela-kotulska-frequenz replied:
done

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py:
```diff
@@ -65,39 +69,106 @@ def generate(
                 )
             )
-            for bat_id in component_ids
-        )
-
-        if not all(battery_inverters):
-            raise ComponentNotFound(
-                "All batteries must have at least one inverter as a predecessor."
-            )
+            if len(inverters) == 0:
+                raise ComponentNotFound(
+                    "All batteries must have at least one inverter as a predecessor.",
+                    "Battery ID %s has no inverter as a predecessor.",
+                    bat_id,
+                )
```
It is interesting that we need this type of checks in the code.  I'd have thought this is already done by the component graph.

Anyway, this is not a log statement, so the `%s` won't be replaced, you need to use an f-string.

ela-kotulska-frequenz replied:
> Anyway, this is not a log statement, so the %s won't be replaced, you need to use an f-string.

Done

> It is interesting that we need this type of checks in the code. I'd have thought this is already done by the component graph.

I didn't check if component graph has it.

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py:
```diff
@@ -65,39 +69,106 @@ def generate(
+        # all_connected_batteries has to be set because
+        # multiple inverters can be connected to the same battery
+        inv_bat_mapping: dict[Component, set[Component]] = {}
+        all_connected_batteries: set[Component] = set()
```
`all_connected_batteries` appears to have been a per-battery variable, to identify all the sibling batteries to any given battery.  Maybe it was not clear from the name.

It shouldn't be moved outside of the `for bat_id in component_ids:` loop, because it does not serve its purpose in that level.

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py:
```diff
@@ -65,39 +69,106 @@ def generate(
                 )
+                all_connected_batteries.update(inv_bat_mapping[inverter])
 
         if len(all_connected_batteries) != len(component_ids):
             raise FormulaGenerationError(
                 "All batteries behind a set of inverters must be requested."
             )
```
This looks wrong as well,  just checking the lengths is not enough, we should check that `all_connected_batteries` is a subset of `component_ids`.

But because this is an older issue, we can do it separately if you don't want to touch it now.

ela-kotulska-frequenz replied:
Fixed

shsms commented on tests/timeseries/_battery_pool/test_battery_pool.py:
```diff
@@ -515,21 +551,43 @@ async def test_batter_pool_power_two_batteries_per_inverter(
                 ),
-                ComponentCategory.METER,
                 (
-                    ComponentCategory.INVERTER,
-                    [ComponentCategory.BATTERY, ComponentCategory.BATTERY],
+                    ComponentCategory.METER,
+                    (
+                        ComponentCategory.INVERTER,
+                        [ComponentCategory.BATTERY, ComponentCategory.BATTERY],
+                    ),
```
Inverters being connected to the main meter is a case we have to support.  If we want to remove it, it has to be discussed first.  Maybe you can bring it up on Friday.

shsms replied:
Oops, never mind, I see the original graph was wrong.

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_consumer_power_formula.py:
```diff
@@ -27,6 +30,25 @@ class ConsumerPowerFormula(FormulaGenerator[Power]):
+
+        Returns:
+            True if the grid component has a meter mounted, False otherwise.
+        """
```
I don't understand the name of the function or the docstrings.  What does `mounted` mean for a meter?  I'd call it `are_grid_meters`.

Separately, based on the outcomes on Friday,  we might have to move this to a more generic place for example in the component graph, so it doesn't have to be dealt with separately everywhere.

ela-kotulska-frequenz replied:
I did it before our conversation. 

>  we might have to move this to a more generic place for example in the component graph, so it doesn't have to be dealt with separately everywhere.

Agree. IS it ok if I leave it like this until Friday?

ela-kotulska-frequenz replied:
> I'd call it are_grid_meters.

Done

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_simple_power_formula.py:
```diff
@@ -0,0 +1,69 @@
+
+
+class SimplePowerFormula(FormulaGenerator[Power]):
+    """Formula generator from component graph for calculating sum of Power.
+
+    The producer power is calculated by summing up the power of requested components.
+
+    Raises:
+        RuntimeError: If no components are defined in the config or if any
+            component is not found in the component graph.
+    """
```
Not sure if the second line belongs here.

```suggestion
    """Formula generator from component graph for calculating sum of Power.

    Raises:
        RuntimeError: If no components are defined in the config or if any
            component is not found in the component graph.
    """

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_grid_power_formula.py:
```diff
@@ -30,6 +36,26 @@ def generate(  # noqa: DOC502
         )
         grid_successors = self._get_grid_component_successors()
 
+        # Ensure the device has an `ACTIVE_POWER` metric.  When inverters
+        # produce `None` samples, those inverters are excluded from the
+        # calculation by treating their `None` values as `0`s.
+        #
+        # This is not possible for Meters, so when they produce `None`
+        # values, those values get propagated as the output.
+        components = {
+            c
+            for c in grid_successors
+            if c.category
+            in {
+                ComponentCategory.INVERTER,
+                ComponentCategory.EV_CHARGER,
+                ComponentCategory.METER,
+            }
+        }
```
I think that comment can go away.  It doesn't make sense after your refactor.

ela-kotulska-frequenz replied:
Done

ela-kotulska-frequenz commented:
@llucax 

>  I find it confusing that we have this extra inheritance FallbackFormulaMetricFetcher -> FallbackMetricFetcher. Can't we get rid of FallbackMetricFetcher? What is it for? We already have some unnecessary inheritance in the SDK from the past, I would prefer not to introduce more unless it is really necessary.

We need it because of circular imports :(
**We would have:**
FormulaGenerator -> FormulaEngine -> MetricFetcher -> FallbackFormulaMetricFetcher -> FormulaGenerator

**Now we have**
FormulaGenerator -> FallbackFormulaMetricFetcher -> FallbackMetricFetcher
FormulaGenerator -> FormulaEngine -> MetricFetcher -> FallbackMetricFetcher

https://github.com/frequenz-floss/frequenz-sdk-python/pull/1037#discussion_r1723439467

I added comment to the FallbackFormulaMetricFetcher:
https://github.com/frequenz-floss/frequenz-sdk-python/pull/1037/commits/23299b4ac212b473836ca43c34eacc678758732f#diff-9800776f7bac0eccbc7d0779b356f024334b154255b892b353dd85c69201f1bfR15

ela-kotulska-frequenz commented:
FallbackMetricFetcher base class is Receiver as @llucax  suggested.

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_fallback_formula_metric_fetcher.py:
```diff
@@ -0,0 +1,98 @@
+        # That's why store it in `_latest_sample`, too.
+        self._sample = await self._receiver.receive()
+        self._latest_sample = self._sample
+        return True
```
Did you try to just forward the call to the underlying receiver instead of re-implementing the receiver logic? I mean:
```suggestion
        return await self._receiver.ready()
```

ela-kotulska-frequenz replied:
Right, done :)

llucax commented on src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_fallback_formula_metric_fetcher.py:
```diff
@@ -0,0 +1,98 @@
+
+        sample = self._sample
+        self._sample = None
+        return sample
```
And:
```suggestion
        return self._receiver.consume()
```

ela-kotulska-frequenz replied:
done

llucax commented:
> We need it because of circular imports :( **We would have:** FormulaGenerator -> FormulaEngine -> MetricFetcher -> FallbackFormulaMetricFetcher -> FormulaGenerator
> 
> **Now we have** FormulaGenerator -> FallbackFormulaMetricFetcher -> FallbackMetricFetcher FormulaGenerator -> FormulaEngine -> MetricFetcher -> FallbackMetricFetcher
> 
> [#1037 (comment)](https://github.com/frequenz-floss/frequenz-sdk-python/pull/1037#discussion_r1723439467)
> 
> I added comment to the FallbackFormulaMetricFetcher: [23299b4#diff-9800776f7bac0eccbc7d0779b356f024334b154255b892b353dd85c69201f1bfR15](https://github.com/frequenz-floss/frequenz-sdk-python/commit/23299b4ac212b473836ca43c34eacc678758732f#diff-9800776f7bac0eccbc7d0779b356f024334b154255b892b353dd85c69201f1bfR15)

I see :cry:

shsms commented:
The switch to `Receiver` interface for `FallbackMetricFetcher` has made it much simpler and opens the door for more future cleanup.

I'd like to see two more changes, can be in this PR or can be turned into an issue:

1. I feel the `latest_sample` field in `FallbackFormulaMetricFetcher` doesn't belong there.  It took me a while to figure out how it was used in the `MetricFetcher`.

    Instead, I'd like to see a variable `_latest_fallback_sample` in the `MetricFetcher` itself, because that's where it is used.  So keeping it entirely there would make it easier to follow when it gets set as well as when it gets read.  And the `FallbackMetricFetcher.latest_sample` field can go away.

2. Then we can rename the `FallbackMetricFetcher` abc to `LazyReceiver` and move it to `_internal._channels`.  And `FallbackFormulaMetricFetcher` can be renamed to `FallbackFormulaMetricReceiver`.

NB: if you're doing it in this PR, please put in a new commit, so it is easier to read the changes.

shsms reviewed (dismissed):
This is great, one more really cool SDK feature!

Approving from my side, but please wait for @llucax's approval also.

ela-kotulska-frequenz committed:
Move latest_sample from FallbackFormulaMetricFetcher to MetricFetcher

Because that's where it is used.

Signed-off-by: Elzbieta Kotulska <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz commented:
@sahas-subramanian-frequenz 

> I feel the latest_sample field in FallbackFormulaMetricFetcher doesn't belong there. It took me a while to figure out how it was used in the MetricFetcher.

Done (last commit)

>  Then we can rename the FallbackMetricFetcher abc to LazyReceiver and move it to _internal._channels. And FallbackFormulaMetricFetcher can be renamed to FallbackFormulaMetricReceiver.

In separate PR: https://github.com/frequenz-floss/frequenz-sdk-python/issues/1056

shsms commented:
> @sahas-subramanian-frequenz

I don't use that account anymore, so I won't be notified if you use it.

The latest commit looks good as well.

llucax reviewed (approved):
> This is great, one more really cool SDK feature!

:heavy_plus_sign: 1

Thanks for taking care of it!
